### PR TITLE
Add CodeBlockTimer and CodeBlockMultipleTimer.

### PR DESCRIPTION
**Pull request by bolotovskyr**
**Purpose of work**

To quickly profile a block of code it's useful to have a light-weight timer class. The existing `Mantid` `Timer` class requires that we instantiate a timer before the code block execution begins and then update its status (i.e. ask to calculate and report the elapsed time) right after the code execution ends.  That might be inconvenient or even not work in certain scenarios. For example, a function being profiled might have multiple `return` points. So if we declare a `Timer` object at the start of that function, we need to place calls to our timer before every `return` statement. Also, if the function throws an exception, the timer might go out of scope before it can be called. This PR proposes a more versatile approach to timing a C++ code block.

**Description of work**

Two new classes are proposed: `CodeBlockTimer` and `CodeBlockMultipleTimer`. These classes are based on the RAII technique: they measure the life time of a timer object.  The timer needs to be instantiated on the stack at the beginning of the code block. The timer constructor saves the start time. When the timer goes out of scope, the destructor calculates and reports the elapsed time. The difference between the two timer classes is in how they report the elapsed time. `CodeBlockTimer` outputs the elapsed time into a stream supplied during the timer construction. This could be, for example, `std::cout` or a `Mantid Logger` stream. `CodeBlockMultipleTimer` is designed to measure the elapsed time in a code block that is executed multiple times. This class works in conjunction with `CodeBlockMultipleTimer::TimeAccumulator` class. A time accumulator object must be declared outside of the code block being timed and passed into the `CodeBlockMultipleTimer` constructor. (One possibility is to declare an accumulator object with a static linkage in the same compilation unit and provide an accessor method.) Every time a `CodeBlockMultipleTimer` object goes out of scope, its destructor updates the time accumulator. When the code block is no longer visited, the time accumulator can be accessed to output the total timing results. One additional benefit of using `CodeBlockMultipleTimer` is that it reports how many times a block of code was executed, which might be useful to know.

**How to use CodeBlockTimer**
```cpp
#input "MantidKernel/Timer.h"
void MyAlgorithm::doSomething()
{
   Mantid::Kernel::CodeBlockTimer timer("MyAlgorithm::doSomething", std::cout);
   // Do something that takes 5 sec
   // Timer will output the elapsed time when the object is destroyed
}

Output: Elapsed time (sec) in "MyAlgorithm::doSomething": 5 
```

**How to use CodeBlockMultipleTimer**
```cpp
FILE: MyAlgorithm_A.cpp

#input "MantidKernel/Timer.h"
// Declare a time accumulator with a static linkage
static Mantid::Kernel::CodeBlockMultipleTimer::TimeAccumulator s_timeAccumulator("MyAlgorithm_A::doSomething");
// Provide an accessor function
MANTID_KERNEL_DLL const Mantid::Kernel::CodeBlockMultipleTimer::TimeAccumulator& myAlgorithm_A_TimeAccumulator(){
  return s_timeAccumulator;
}
...
void MyAlgorithm_A::doSomething()
{
   Mantid::Kernel::CodeBlockMultipleTimer timer(s_timeAccumulator);
   // Do something that takes 5 sec
   // Timer will update TimeAccumulator when the object is destroyed 
}
-----------------------------
File: MyAlgorithm_B.cpp
#include "Timer.h"
// Let the linker know that the accessor function for the time accumulator is defined in another compilation unit
extern const Mantid::Kernel::CodeBlockMultipleTimer::TimeAccumulator& myAlgorithm_A_TimeAccumulator();

void MyAlgorithm_B::doSomething(){
   // Call MyAlgorithm_A::doSomething() 3 times
   // ...
   // Output the results
   std::cout << myAlgorithm_A_TimeAccumulator() << '\n';
}

Output: "Elapsed time (sec) in "MyAlgorithm_A::doSomething":  15; Number of entrances: 3
```

**To test:**
`ctest -R KernelTest_TimerTest`

_There is no associated issue._

_This does not require release notes_ because it will never be directly observed by a user.

### CODE REVIEW

**Review by peterfpeterson — COMMENTED** [review]
Please add a method 
```cpp
MANTID_KERNEL_DLL std::ostream &operator<<(std::ostream &, const CodeBlockTimer &);
```
to make this easier to integrate into people's usage

File: Framework/Kernel/inc/MantidKernel/Timer.h [hunk 1]
```diff
@@ -40,5 +40,47 @@ class MANTID_KERNEL_DLL Timer {
...
+
+  public:
+    void reset();
+    void increment(const double time_s);
+    double getElapsed() const;
+    size_t getNumberOfEntrances() const;
+    void report(std::ostream &out) const;
+
+  private:
+    std::string name;
```
Comment (peterfpeterson) [inline Framework/Kernel/inc/MantidKernel/Timer.h:70]: According to the [mantid c++ coding standards](https://developer.mantidproject.org/Standards/CPPStandards.html#naming-conventions), these should be `m_name`, `m_elapsed_sec`, and `m_number_of_entrances`. I added the extra characters to the elapsed time to make it slightly more explicit.

File: Framework/Kernel/src/Timer.cpp [hunk 2]
```diff
@@ -60,4 +60,70 @@ std::ostream &operator<<(std::ostream &out, const Timer &obj) {
...
+ */
+void CodeBlockMultipleTimer::TimeAccumulator::reset() {
+  elapsed_s = 0.0;
+  number_of_entrances = 0;
+}
+
+/** Increment the elapsed wall-clock time and number of times the code block was entered
+ * @param time_s :: elapsed time (seconds) to add
+ */
+void CodeBlockMultipleTimer::TimeAccumulator::increment(const double time_s) {
```
Comment (peterfpeterson) [inline Framework/Kernel/src/Timer.cpp:110]: I haven't seen usage, but would it simplify clients by adding an overload that takes two timestamps (i.e. `std::chrono::system_clock::now()`)?

File: Framework/Kernel/inc/MantidKernel/Timer.h (hunk 1 above)
Comment (bolotovskyr) [reply Framework/Kernel/inc/MantidKernel/Timer.h]: I have updated all class member names and changed `s` to `sec`.

File: Framework/Kernel/src/Timer.cpp (hunk 2 above)
Comment (bolotovskyr) [reply Framework/Kernel/src/Timer.cpp]: It is true that, in principle, time calculation may be implemented in either `CodeBlockMultipleTimer` or `CodeBlockMultipleTimer::TimeAccumulator`. Still, I think it's a cleaner design when the timer calculates the elapsed time after each code block execution, and the time accumulator only keeps track of the total time spent in all calls.

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**